src/Codegen/Constraints/BaseBuilder.php

```diff
@@ -92,6 +93,64 @@ protected function addEnumConstraintCheck(HackBuilder $hb): void {
     }
   }
 
+  protected function addHackEnumConstraintCheck(HackBuilder $hb): void {
+    $schema = type_assert_type($this->typed_schema, TSchema::class);
+    if (Shapes::keyExists($schema, 'hackEnum')) {
+      try {
+        $rc = new \ReflectionClass($schema['hackEnum']);
+      } catch (\ReflectionException $e) {
+        throw new \Exception(Str\format("Hack enum '%s' does not exist", $schema['hackEnum']));
+      }
+
+      invariant($rc->isEnum(), "'%s' is not an enum", $schema['hackEnum']);
+
+      $schema_type = $schema['type'] ?? null;
+      $hack_enum_values = keyset[];
+      foreach ($rc->getConstants() as $hack_enum_value) {
+        if ($schema_type === TSchemaType::INTEGER_T) {
+          $hack_enum_value = $hack_enum_value ?as int;
+        } else {
+          $hack_enum_value = $hack_enum_value ?as string;
+        }
+        invariant(
+          $hack_enum_value is nonnull,
+          "'%s' must contain only values of type %s",
+          $rc->getName(),
+          $schema_type === TSchemaType::INTEGER_T ? 'int' : 'string'
+        );
+        $hack_enum_values[] = $hack_enum_value;
+      }
+
+      if (Shapes::keyExists($schema, 'enum')) {
+        // If both `enum` and `hackEnum` are specified, assert that `enum` is a subset of
+        // `hackEnum` values.
+        foreach ($schema['enum'] as $enum_value) {
+          invariant(
+            $enum_value is string,
+            "Enum value '%s' is not a valid value for '%s'",
+            \print_r($enum_value, true),
+            $rc->getName()
+          );
+          invariant(
+            C\contains_key($hack_enum_values, $enum_value),
+            "Enum value '%s' is unexpectedly not present in '%s'",
+            \print_r($enum_value, true),
+            $rc->getName()
+          );
+        }
+      }
```

Comment:
I read and re-read the PR description but it is still not clear to me why we check on `enum` as well. It it because there are other constraint checks against `string` and `integer` schema types that rely on `hack` keys?

Reply:
It's because we also check that the provided value is one of the values specified by `enum`, so if we provide a value in `enum` which is not also present in `hackEnum` the generator would never accept that value. So permitting the value is pointless.

We could also enforce that all values present in `hackEnum` are present in `enum`, but that feels a bit too constraining given we might want to use some value in the Hack enum elsewhere in Hack code.